Libplanet.Explorer/Queries/StateQuery.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Security.Cryptography;
+using GraphQL;
+using GraphQL.Types;
+using Libplanet.Action;
+using Libplanet.Assets;
+using Libplanet.Blockchain;
+using Libplanet.Blockchain.Policies;
+using Libplanet.Blocks;
+using Libplanet.Explorer.GraphTypes;
+
+namespace Libplanet.Explorer.Queries;
+
+public class StateQuery<T>
+    : ObjectGraphType<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)>
+    where T : IAction, new()
+{
+    public StateQuery()
+    {
+        Name = "StateQuery";
+        Field<NonNullGraphType<FungibleAssetValueType>>(
+            "balance",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<AddressType>> { Name = "owner" },
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveBalance
+        );
+        Field<FungibleAssetValueType>(
+            "totalSupply",
+            arguments: new QueryArguments(
+                new QueryArgument<NonNullGraphType<ByteStringType>> { Name = "currencyHash" },
+                new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "offsetBlockHash" }
+            ),
+            resolve: ResolveTotalSupply
+        );
+    }
+
+    private static Currency GetNativeTokenFromHash(
+        IResolveFieldContext<(IBlockChainStates<T> ChainStates, IBlockPolicy<T> Policy)> context,
+        string name
```

Comment:
Would it matter for this parameter to have a more descriptive name?

Reply:
I'm going to rename it to `paramName`.

Reply:
Amended.